airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -309,15 +308,34 @@ def post_set_task_instances_state(*, dag_id: str, session: Session = NEW_SESSION
         error_message = f"Task ID {task_id} not found"
         raise NotFound(error_message)
 
-    execution_date = data['execution_date']
-    try:
-        session.query(TI).filter_by(execution_date=execution_date, task_id=task_id, dag_id=dag_id).one()
-    except NoResultFound:
-        raise NotFound(f"Task instance not found for task {task_id} on execution_date {execution_date}")
+    execution_date = data.get('execution_date')
+    run_id = data.get('dag_run_id')
+
+    if run_id is not None:
+        ti = (
+            session.query(func.count(TI.dag_id))
+            .filter(TI.task_id == task_id, TI.dag_id == dag_id, TI.run_id == run_id)
+            .scalar()
+        )
+        if not ti:
+            raise NotFound(
+                detail=f"Task instance not found for task {task_id!r} on DAG run with ID {run_id!r}"
+            )
+    elif execution_date is not None:
+        ti = (
+            session.query(TI.run_id)
+            .filter(TI.execution_date == execution_date, TI.task_id == task_id, TI.dag_id == dag_id)
+            .first()
+        )
+        if not ti:
+            raise NotFound(
+                detail=f"Task instance not found for task {task_id!r} on execution_date {execution_date}"
+            )
+        run_id = ti.run_id
```

Comment:
Can (should?) this backfilling run_id from execution_date be done in the schema class?

Reply:
Have checked on this, we don't have access to the ` dag` at the schema level. Looking up the task instance may not always give a correct result

Reply:
One thing is, we don't also need to backfill it because `dag.set_task_instance_state` also accepts execution date. So I will work on passing the execution_date instead of the backfill

Reply:
Makes sense. Do you think it’d be worthwhile to refactor this out as a function for reuse (is it useful?)

Reply:
Yeah. I have refactored it into taskinstance so we can now do TaskInstance.find just like we do for DagRun